search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,517 @@
+  GDBusMethodInvocation      *invocation;
+} ArticleMetadataQueryState;
+
+static ArticleMetadataQueryState *
+article_metadata_query_state_new (EksArticleMetadataProvider *provider,
+                                  GDBusMethodInvocation      *invocation)
+{
+  ArticleMetadataQueryState *state = g_new0 (ArticleMetadataQueryState, 1);
+  state->provider = provider;
+  state->invocation = invocation;
+
+  return state;
+}
+
+static void
+article_metadata_query_state_free (ArticleMetadataQueryState *state)
+{
+  g_free (state);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ArticleMetadataQueryState,
+                               article_metadata_query_state_free);
+
+static gboolean
+models_for_result (EkncEngine   *engine,
+                   const gchar  *application_id,
+                   GAsyncResult *result,
+                   GSList       **models,
+                   GError       **error)
+{
+  g_autoptr(EkncQueryResults) results = NULL;
+  if (!(results = eknc_engine_query_finish (engine, result, error)))
+      return FALSE;
+
+  EkncDomain *domain = eknc_engine_get_domain_for_app (engine,
+                                                       application_id,
+                                                       error);
+  if (domain == NULL)
+      return FALSE;
+
+  *models = g_slist_copy_deep (eknc_query_results_get_models (results),
+                               (GCopyFunc) g_object_ref,
+                               NULL);
+
+  return TRUE;
+}
+
+static gboolean
+models_and_shards_for_result (EkncEngine   *engine,
+                              const gchar  *application_id,
+                              GAsyncResult *result,
+                              GSList       **models,
+                              GSList       **shards,
+                              GError       **error)
+{
+  g_autoptr(EkncQueryResults) results = NULL;
+  if (!(results = eknc_engine_query_finish (engine, result, error)))
+      return FALSE;
+
+  EkncDomain *domain = eknc_engine_get_domain_for_app (engine,
+                                                       application_id,
+                                                       error);
+  if (domain == NULL)
+      return FALSE;
+
+  *shards = g_slist_copy_deep (eknc_domain_get_shards (domain),
+                               (GCopyFunc) g_object_ref,
+                               NULL);
+  *models = g_slist_copy_deep (eknc_query_results_get_models (results),
+                               (GCopyFunc) g_object_ref,
+                               NULL);
+
+  return TRUE;
+}
+
+static GStrv
+strv_from_shard_list (GSList *string_list)
+{
+  GStrv strv = g_new0 (gchar *, g_slist_length (string_list));
+  guint count = 0;
+
+  for (GSList *l = string_list; l; l = l->next)
+    {
+      EosShardShardFile *shard = l->data;
+      gchar  *shard_path = NULL;
+
+      g_object_get (shard, "path", &shard_path, NULL);
+      strv[count++] = shard_path;
+    }
+
+  return strv;
+}
```

Comment:
Its probably worth creating a single translation unit for both the discovery feed provider and this with these functions